### DISCUSSION TITLE
ZSTD

### DISCUSSION THREAD

**Discussion by rsmith22** [libvips/libvips#3469] [Q&A]
Can you tell me whether libvips support for ZSTD using libtiff can convert a tiff image compressed with LZW to a tiff image compressed with ZSTD?

**Comment by jcupitt** [comment]
Sure, just:

```
vips copy x.tif y.tif[compression=zstd]
```

**Reply by rsmith22** [reply]
hello and thank you for your input
I built libtiff with ZSTD, so when I convert from LZW to ZSTD there is problem with the resulting image it is made up of lower byte (0-256) and the highest bit. The image has lost much of the greyscale and looks binary. The horizontal predicator for ZSTD was set the same from LZW.

**Reply by jcupitt** [reply]
Sorry, I don't know what might be wrong. You'd need to give a lot more information.

I tried in ubuntu 23.04 and 22.04 and it seems to work fine:

```
$ vips copy astronauts.png zstd.tif[compression=zstd]
$ tiffinfo zstd.tif 
=== TIFF directory 0 ===
TIFF Directory at offset 0x11763a (1144378)
  Image Width: 900 Image Length: 900
  Resolution: 10, 10 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: ZSTD
  Photometric Interpretation: RGB color
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 3
  Rows/Strip: 128
  Planar Configuration: single image plane
  Predictor: horizontal differencing 2 (0x2)

$ vips copy zstd.tif x.png
$
```

And it looks OK:

![x](https://user-images.githubusercontent.com/580843/234916934-e3b93e74-c63f-4f79-8f3e-e650dca83af3.png)

**Comment by rsmith22** [comment]
![image](https://user-images.githubusercontent.com/101135353/234919723-ca4c89c4-d776-4537-8ecf-b1bdcd3476ed.png)

**Comment by rsmith22** [comment]
The code is reading / writing tiles, should it be Rows/Strip for ZSTD?

**Comment by rsmith22** [comment]
Could I upload the tiff LZW image for you to convert to tiff ZSTD?

**Reply by jcupitt** [reply]
Sure, or you could install libvips in a VM perhaps. What platform are you using?

**Comment by rsmith22** [comment]
I cannot attach the tif file compressed with LZW, so it's saved as JPEG. It will not surprise me if this jpeg converts to ZSTD. 

The OS is Windows.

The libvips version I downloaded was 8.10 and should support ZSTD. However, when I initially used it to convert to ZSTD it resulted in an error stating "ZSTD is not configured". I assumed libtiff was not built to support ZSTD. Are there binaries I can download with ZSTD supported?
![SWTEST3](https://user-images.githubusercontent.com/101135353/235185400-8bf14431-a83b-42bd-807e-b3a2fd02a68b.jpg)

**Reply by jcupitt** [reply]
Ah OK, yes the win binaries don't include zstd, you'd need to build libtiff yourself for that.

You could use WSL and then `apt install libvips-tools` there.

**Comment by rsmith22** [comment]
Yes, when I built libtiff with ZSTD the conversion resulted in the image above.
Not sure what WSL is, and if I were to install the libvips tools does this include a libtiff binary that supports ZSTD?

**Reply by jcupitt** [reply]
https://learn.microsoft.com/en-us/windows/wsl/install

It's a very simple way to run linux inside windows.

**Comment by rsmith22** [comment]
hmmm, binaries built to run in linux are not going to help my development. Any suggestion how I can confirm libtiff with ZSTD support will convert LZW tiff to ZSTD tiff with binaries that run in Windows?

**Reply by jcupitt** [reply]
Right, but WSL will let you run those binaries in windows. 

It's just two commands -- first, run this in an admin powershell window:

```
wsl --install
```

Then open a WSL terminal and enter:

```
apt install libvips-tools
```

And now you can run eg:

```
vips copy x.tif y.tif[compression=zstd]
```

**Reply by rsmith22** [reply]
yes, I'll try and let you know. However, it would appear that I still have to overcome the issue of building libtiff with ZSTD support that works in Windows

**Reply by jcupitt** [reply]
I'd try adding zstd to the mxe build system:

https://github.com/libvips/build-win64-mxe

**Reply by rsmith22** [reply]
thank you
